### PR TITLE
[otbn,doc] Document the errors that the various instructions can cause

### PR DESCRIPTION
**Pull request by rswarbrick**
The first 3 commits are just clean-ups to how we generate documentation. Then commit 4 adds a field to the YAML schema and teaches the ISA documentation to show it. Finally, commit 5 adds the relevant information to all the different instructions.

The result looks like this for an instruction that has interesting error behaviour:
image

And this for one that doesn't:
image

@GregAC: Would you mind scanning through the list of possible errors and checking that it matches what you expect? You might find it easier to do this in the YAML file, where we use &anchors: I found my eyes glazing over a little less that way!

This is related to https://github.com/lowRISC/opentitan/issues/7809 because it makes it much easier to spot the instructions where it's possible to have two things go wrong at once. My shortlist for that is:

- LW
- SW
- BEQ
- BNE
- JAL
- JALR
- CSRRS
- CSRRW
- LOOP
- LOOPI
- BN.LID
- BN.SID
- BN.MOVR

(which is shorter than I feared, but longer than I hoped!)

### CODE REVIEW

File: hw/ip/otbn/data/base-insns.yml
Comment (imphil) [inline hw/ip/otbn/data/base-insns.yml:543]:
```suggestion
    - A `LOOP` error if `iterations` is zero.
```

File: hw/ip/otbn/data/base-insns.yml
Comment (imphil) [inline hw/ip/otbn/data/base-insns.yml:342]:
```suggestion
    - A `BAD_INSN_ADDR` error if the branch is taken and the computed address is not a valid PC.
```

File: hw/ip/otbn/data/bignum-insns.yml
Comment (imphil) [inline hw/ip/otbn/data/bignum-insns.yml:688]: That's equal to the "grs12-call-stack" anchor (if you want).

File: hw/ip/otbn/util/yaml_to_doc.py
Comment (imphil) [inline hw/ip/otbn/util/yaml_to_doc.py:281]: Two comments regarding the wording here:

- We don't use or define the term "synchronous errors" anywhere in the spec.
- The errors listed here are only the ones that are instruction-related/software-related. There might be other errors which are happening when accessing resources during the execution of the instruction, e.g. an integrity error when reading the register file or the data memory.

I left a note at https://github.com/lowRISC/opentitan/pull/8243#issuecomment-921851464 to discuss the terminology, but maybe the following change will do the trick:

```suggestion
            parts.append('{} might cause the following software errors:\n'
                         .format(mnem))
```

**Review by imphil — APPROVED** [review]
Thanks @rswarbrick! As expected, a couple wording comments from my side :smile: 

Thinking ahead, there might be value in specifying the error code and the description of what causes the error separately, e.g. to extract a list of valid error combinations out of it and convert it into an assert. But that can easily come in a follow-up when we actually need it.

```
errs:
  - name: CALL_STACK
    condition: A `CALL_STACK` error from using `x1` as `grs1` or `grs2` when the call stack is empty.
```

File: hw/ip/otbn/data/bignum-insns.yml
Comment (rswarbrick) [reply hw/ip/otbn/data/bignum-insns.yml:688]: Ahah, but that's defined in a different file :-) I tried that at first and took a few seconds to work out why the YAML parsing was failing!

File: hw/ip/otbn/util/yaml_to_doc.py
Comment (rswarbrick) [reply hw/ip/otbn/util/yaml_to_doc.py:281]: Sounds like a good idea for now, at least.

**Review by GregAC — APPROVED** [review]
LGTM, though do we want to mention fatal errors (specifically fatal_dmem, fatal_reg)?

**Comment by imphil** [comment]
> LGTM, though do we want to mention fatal errors (specifically fatal_dmem, fatal_reg)?

I'd actually be inclined to mark these integrity violation errors as "can happen at any time", since their detection isn't necessarily a property of the ISA or the software code, but more of how and when we check for integrity violations.

But then we actually know when we check for integrity violations (on every access), so we could use this information in verification. If we want to do that, we'd need to go with your suggestion Greg and add the fatal errors as well (or just have a blank statement: all instructions accessing registers can raise fatal_reg, all instructions accessing instruction memory can raise fatal_imem, and all instructions can raise fatal_imem).

I'm personally happy either way.

**Comment by GregAC** [comment]
>  since their detection isn't necessarily a property of the ISA or the software code, but more of how and when we check for integrity violations.

I think architecturally defining when integrity checks will occur is sensible.

A blanket statement (all instructions that access registers may cause FATAL_REG) would do, though for the call stack we're listing the error on all instructions that access base registers so we should probably do the same for FATAL_REG/FATAL_DMEM. Though this is a little awkward for the 'synchronous' -> 'software' terminology switch as FATAL_REG/FATAL_DMEM wouldn't be software errors.

I think a blank statement is fine for FATAL_IMEM as there's no associated instruction (it's similar to ILLEGAL_INSN for when an instruction has an invalid encoding).

**Comment by rswarbrick** [comment]
Thanks for the comments. I'll add a general note at the top, since it seems a bit silly to write something on every instruction except ECALL and LOOPI! :-) There's a related question about how to do DV for this, but I guess that belongs elsewhere.

**Comment by rswarbrick** [comment]
Hi @imphil and @GregAC: Are you happy with this (for now, at least)? I've got a PR that hits these coverage points waiting and I'd rather merge this before pushing it up if it's ready for merge anyway.

**Pull request merged**
